0000-experiment-dataframe.md

```diff
@@ -0,0 +1,444 @@
+
+- `IBMExperimentService.create_analysis_results()`
+- `IBMExperimentService.update_analysis_results()`
+- `IBMExperimentService.analysis_result()`
+- `IBMExperimentService.analysis_results()`
+
+## Examples
+
+### A. Writing an analysis class
+
+When an experiment author writes a custom analysis class, one must override the `_run_analysis()` method of a `BaseAnalysis` subclass.
+This style doesn't change by our proposal to minimize the impact to user code.
+Even though the author can directly call the `ExperimentData.add_analysis_results()` and `ExperimentData.add_artifacts()` from the `_run_analysis()`, this pattern must be prohibited because this ties analysis to executor (i.e. we want to make `_run_analysis()` agnostic to the `ExperimentData` in future update).
+
+At present, the `_run_analysis()` method takes the `ExperimentData` as input and returns the `Tuple[List[AnalysisResult], List[Figure]]`. With this proposal, the return type will be generalized with artifact.
+
+```python
+from qiskit_experiments.framework import ExperimentData, AnalysisResult, ArtifactData
+
+class MyAnalysis(BaseAnalysis):
+
+  def _run_analysis(
+    experiment_data: ExperimentData
+  ) -> Tuple[List[AnalysisResult], List[ArtifactData]]:
+
+    # In future, _run_analysis will directly receive this.
+    data = experiment_data.data()
+    ...
+    # Do some analysis. Generate result and figure (and other artifacts).
+    ...
+    result1 = AnalysisResult(
+      name="param1",
+      value=0.1,
+      quality="good",
+      unit="Hz",
+    )
+    artifact1 = ArtifactData(
+      data=mpl_figure,
+      name="curve_plot",
+    )
+    artifact2 = ArtifactData(
+      data=curve_data_table,
+      name="curve_data",
+    )
+    return [result1], [artifact1, artifact2]
+
+```
+
+As you can see, some expected data fields are missing in `result1` in this example (such as component and run time). Note that the minimum required field that the author must provide are name and value. The missing data are implicitly filled by the following base class routine. After this canonicalize, `BaseAnalysis.run` calls `add_analysis_results()` and `add_artifacts()`. The formatted data frame is created (updated) in `ExperimentData` by `add_analysis_results()` call.
+
+### B. Adding new analysis results or artifacts.
+
+An end user (experimentalist) can directly call `add_analysis_results()` and `add_artifacts()` to create custom entry. For example,
+
+```python
+exp_data = MyExperiment(...).run().block_for_result()
+
+exp_data.add_analysis_results(
+  name="my_param1",
+  value="0.1",
+)
+exp_data.add_artifacts(
+  name="my_second_plot",
+  data=my_plot2,
+)
+```
+
+In this case, the user can bypass creation of `AnalysisResult` or `ArtifactData` since these are assumed to be a container used by an experiment author to write `_run_analysis`.
+
+### C. Updating existing results or artifacts.
+
+An end user (experimentalist) can modify analysis result or artifact after experiment.
```

Comment:
Is this intended to replace the artifact with a new one using the same `artifact_id` or to delete the existing one and add the replacement artifact with a new `artifact_id`?

Reply:
I think it depends on the experiment service implementation, but I was thinking to replace existing entry with the same ID.